Virtual scroll integration

dhilt opened the pull request:
This is for https://github.com/dhilt/ngx-ui-scroll/issues/306.

dhilt committed:
Do not create a new message when the status of the old message is changed. Replace contents instead.

dhilt committed:
Single datasource: append/replace instead of reset

dhilt committed:
Avoid float and margin

dhilt committed:
Virtualization during append

dhilt committed:
Cleanup

dhilt commented on src/app/chat/chat-message/chat-message.component.scss:
```diff
@@ -24,7 +24,7 @@ div {
   &.external-message {
     background-color: #d6d6d6;
-    float: left;
+    // float: left;
```
Scroller doesn't like [margin](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing) and float; its inner calculations are very minimalistic and don't cover all the possible cases that can be implemented as the CSS level. The best way would be to use another approach there (like padding instead of margin).

fabioemoutinho replied:
thanks for the tip!

dhilt commented on src/app/chat/chat.component.ts:
```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
+        .pipe(take(1), tap(console.log));
+    },
+    devSettings: { debug: true },
+  });
```
Since the update is not total, the Datasource should remain. But we need to provide merge logic, see below.

dhilt commented on src/app/chat/chat.component.ts:
```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
+            this.replaceMessage(oldItem);
+            return;
+          }
+        })
```
Here's the core change. By prev/curr arrays we can detect change and make a decision: add new item or update the existed one. New item is detected by new id. Item to update is detected by time change. I had to reconsider the id logic at the effects level, see below.

fabioemoutinho replied:
interesting. I would love to avoid nested iterations though, if possible.
I'm thinking I could trigger new or old message updates at the store effects, calling a service that holds the datasource.

dhilt replied:
The Scroller doesn't know if there are any changes in the data before it requests this data via datasource.get method. So if the changing items are present in the DOM (say, not virtual), we need to think how to tell the Scroller about it. And the adapter methods are designed to provide such changes in the runtime.

dhilt commented on src/app/chat/chat.component.ts:
```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
+    await this.datasource.adapter.clip({
+      forwardOnly: true,
+    });
+  }
```
So, we use Adapter.append to push new item to the end of the list on the UI. But if the end-of-file is not reached, this push will be virtual (`eof: true` provides such behaviour). Also, we are clipping items after append to provide virtualization over unlimited appending. Clip will break through the end-of-file at some point, and new items will be virtualized.

dhilt commented on src/app/chat/chat.component.ts:
```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
+    await this.datasource.adapter.relax();
+    await this.datasource.adapter.replace({
+      predicate: ({ data }) => data.id === message.id,
+      items: [message],
```
This provides hard replace: item is being removed and re-created again with new contents. In case of replacing virtual items this doesn't affect the UI. But if it replaces the in-buffer items, I would say, there may be a place for optimization, and perhaps you may want to look at the [adapter-fix-updater](https://dhilt.github.io/ngx-ui-scroll/#/experimental#adapter-fix-updater) API...

fabioemoutinho replied:
I just tested this locally, and once items are added to the DOM (I tried over 30 items) they are never removed. Is this expected? 🤔 I thought only items in the default buffer size (5) would be rendered

fabioemoutinho replied:
if adapter-fix-updater is needed, is there a way to run the updater function only for the item I want? I'm thinking that iterating over the whole collection to update one item might result in poor performance if you have thousands of messages in the collection

dhilt replied:
Predicate's traversing only the in-buffer (non-virtual) items that are present in the DOM, not whole collection. It can't hit performance.

On items remove, I pushed "clip" fix, now it should properly clip in all cases. The issue with the previous revision was that I tested the UI by scrolling to the very bottom each time before a new adding. So if you didn't scroll before new append, it remained unclipped. It is fixed now.

dhilt commented on src/app/store/chat/chat.reducer.ts:
```diff
@@ -52,7 +52,10 @@ export const chatFeature = createFeature({
-          { id: pendingMessageId, map: () => message },
+          {
+            id: pendingMessageId,
+            map: () => ({ ...message, id: pendingMessageId }),
```
This is for not to make new ID when we update the existed item.

fabioemoutinho replied:
this was actually intended behavior. A temporary id is created so the message is added to the collection to appear in the UI before there's a response from the server, but when the response comes back, the server sends the correct ID.

dhilt replied:
No problem with it, you only need to implement one more ID for your case. I mean the UI ID (Scroller's ID), which will persist across the server changes. The Scroller's message is a human-like message remaining the same entity not depending of its contents changes. Status or text (or Server ID) may change, Scroller's ID must not.

dhilt commented on src/app/store/chat/chat.service.ts:
```diff
@@ -130,17 +130,21 @@ export class ChatService {
       body: message.body,
       userId: message.userId,
-      id: crypto.randomUUID(),
+      id: message.id || crypto.randomUUID(),
```
Re-use ID for update case.

fabioemoutinho replied:
I didn't do it this way because even if unlikely to generate conflicting IDs, we shouldn't rely on the front end to generate IDs

fabioemoutinho replied:
the ID generated by the backend might be used later for features like updating or deleting messages

dhilt replied:
Just separate the backend ID and the frontend one. Frontend ID should persist.

dhilt committed:
Clip in both directions

dhilt commented:
@fabioemoutinho I believe we are close to complete this as the core understanding becomes deeper. I would highlight two major points:

 - persisting frontend ID vs mutating backend ID (separation of concerns )
 - syncing of the collection data changes with the Scroller's inner buffer via Adapter API

Also, I imported the project to stackblitz: https://stackblitz.com/edit/angular-ngrx-chat

fabioemoutinho commented:
@dhilt wow the result is great! I'll try to make the changes this week and I will ping you when it's done.
I'll also try to run some performance tests by generating huge amounts of message events and seeing how it performs.

I appreciate you taking the time and effort to help. thank you!

fabioemoutinho commented:
@dhilt I think I've made some progress with performance testing. Could you take a look here? It's generating messages every 50ms and appending them to datasource. I've refactored a bit to put the datasource logic inside a service (virtual-scroll-datasource.service.ts)
https://stackblitz.com/edit/angular-ngrx-chat-c8jbb5?file=src%2Fapp%2Fchat%2Fvirtual-scroll-datasource.service.ts,src%2Fapp%2Fstore%2Fchat%2Fchat.effects.ts,src%2Fapp%2Fstore%2Fchat%2Fchat.service.ts,src%2Fapp%2Fchat%2Fchat.component.ts

I have not been able to find an option to keep the scroll at the bottom. Is there any recommended way to do so?

fabioemoutinho commented:
> @dhilt I think I've made some progress with performance testing. Could you take a look here? It's generating messages every 50ms and appending them to datasource. I've refactored a bit to put the datasource logic inside a service (virtual-scroll-datasource.service.ts) https://stackblitz.com/edit/angular-ngrx-chat-c8jbb5?file=src%2Fapp%2Fchat%2Fvirtual-scroll-datasource.service.ts,src%2Fapp%2Fstore%2Fchat%2Fchat.effects.ts,src%2Fapp%2Fstore%2Fchat%2Fchat.service.ts,src%2Fapp%2Fchat%2Fchat.component.ts
> 
> I have not been able to find an option to keep the scroll at the bottom. Is there any recommended way to do so?

sending messages in an interval less than 50ms starts generating errors. I also had to disable the debug option, or else errors start popping up in the console.

edit:
also, effects are calling the virtual-scroll-datasource.service.ts to append or update messages.

dhilt committed:
mock message pull

dhilt committed:
component fixes, optimization

dhilt committed:
$index consistency flow

dhilt committed:
random height

dhilt committed:
ngx-ui-scroll v3.0.1

dhilt committed:
relative path, stackblitz compatibility

dhilt commented:
@fabioemoutinho I was busy with an important [stability issue](https://github.com/dhilt/vscroll/issues/39) that caused problems while testing your scenario, so the new version (ngx-ui-scroll 3.0.1) should be used, please update!

I got lost in your ngrx changes and, not being able to make my updates on top of them, I decided to update the previous version of the application.

A new serious problem in your approach was highlighted: index consistency. Accessing the data items by natural array indexes (selectMessagesSlice) is broken due to status updates in effects that makes the updating items order wrong, eg: 1, 2, 3, 4, 7, 8, 9, 5, 6. I believe it could be fixed at the natural indexes level, but in order to save time, I introduced a separate $index field, see this commit: https://github.com/fabioemoutinho/angular-ngrx-chat/pull/1/commits/4db5445e9b319d26a5200017320e58e30752db65.

I also added random heights and provided a spam procedure to populate 1000 items. With ngx-ui-scroll v3.0.1 I don't experience issues. https://stackblitz.com/edit/angular-ngrx-chat

On autoscroll after append, there is an old demo: https://stackblitz.com/edit/angular-ngx-ui-scroll-1-2-0-append-autoscroll. But today I would do it as follows:

```
    await this.datasource.adapter.append({ ... });
    this.datasource.adapter.fix({ scrollPosition: Infinity });
```

If you need to scroll to specific item, I would consider this demo: https://dhilt.github.io/ngx-ui-scroll/#/adapter#check-size.

fabioemoutinho commented:
> @fabioemoutinho I was busy with an important [stability issue](https://github.com/dhilt/vscroll/issues/39) that caused problems while testing your scenario, so the new version (ngx-ui-scroll 3.0.1) should be used, please update!
> 
> I got lost in your ngrx changes and, not being able to make my updates on top of them, I decided to update the previous version of the application.
> 
> A new serious problem in your approach was highlighted: index consistency. Accessing the data items by natural array indexes (selectMessagesSlice) is broken due to status updates in effects that makes the updating items order wrong, eg: 1, 2, 3, 4, 7, 8, 9, 5, 6. I believe it could be fixed at the natural indexes level, but in order to save time, I introduced a separate $index field, see this commit: [4db5445](https://github.com/fabioemoutinho/angular-ngrx-chat/commit/4db5445e9b319d26a5200017320e58e30752db65).
> 
> I also added random heights and provided a spam procedure to populate 1000 items. With ngx-ui-scroll v3.0.1 I don't experience issues. https://stackblitz.com/edit/angular-ngrx-chat
> 
> On autoscroll after append, there is an old demo: https://stackblitz.com/edit/angular-ngx-ui-scroll-1-2-0-append-autoscroll. But today I would do it as follows:
> 
> ```
>     await this.datasource.adapter.append({ ... });
>     this.datasource.adapter.fix({ scrollPosition: Infinity });
> ```
> 
> If you need to scroll to specific item, I would consider this demo: https://dhilt.github.io/ngx-ui-scroll/#/adapter#check-size.

oh right, messages are ordered in NgRx by the timestamp that comes back from the server (currently mocked). so it could modify the order when messages are updated, which I'll have to take into account, unfortunately, as it is a requirement I have to follow. I will take some time to look at the changes and test everything, but again I have to thank you for all your effort. 👏
